packages/tenderly-hardhat/src/utils/util.ts

```diff
@@ -1,13 +1,220 @@
+      );
+      return null;
+    }
+
+    contracts.push({
+      contractToVerify: flatContract.name,
+      // TODO(dusan) this can be done with TASK_COMPILE_SOLIDITY_GET_COMPILER_INPUT hardhat task
+      sources: await extractSources(job),
+      compiler: job.getSolcConfig(),
+      networks: {
+        [chainId]: {
+          address: flatContract.address,
+          links: flatContract.libraries,
+        },
+      },
+    });
+  }
+  // TODO(dusan) see about merging compilation jobs?
+
+  return {
+    contracts,
+  };
+}
+
+async function extractSources(job: CompilationJob): Promise<Record<string, TenderlyVerifyContractsSource>> {
+  const sources: Record<string, TenderlyVerifyContractsSource> = {};
+  logger.info("Extracting sources from compilation job.");
+
+  for (const file of job.getResolvedFiles()) {
+    const name = file.sourceName.split("/").slice(-1)[0].split(".")[0];
+    sources[file.sourceName] = {
+      name,
+      code: file.content.rawContent,
+    };
+  }
+
+  return sources;
+}
+
+// TODO(dusan): This function shouldn't be here. Fork verification should have the same request format as private or public verification
+async function _convertToForkRequest(
+  hre: HardhatRuntimeEnvironment,
+  request: TenderlyVerifyContractsRequest
+): Promise<TenderlyForkContractUploadRequest> {
+  const forkRequest: TenderlyForkContractUploadRequest = { config: {}, contracts: [], root: "" };
+
+  // Merge all compilation jobs to see if they can all fit into one request, if not, throw an error because fork verification doesn't support multiple compiler version.
+  const compilationJobs: CompilationJob[] = [];
+  for (const contract of request.contracts) {
+    compilationJobs.push(await getCompilationJob(hre, contract.contractToVerify));
+  }
+  const mergedJob: CompilationJob[] = await hre.run(TASK_COMPILE_SOLIDITY_MERGE_COMPILATION_JOBS, { compilationJobs });
+  if (mergedJob.length > 1) {
+    throw new Error(
+      "The provided contracts must be compiled with the same compiler configuration. At this point, hardhat-tenderly plugin doesn't support multi compiler verification on FORKS. Private and public verification support multi compiler verification."
+    );
+  }
+
+  for (const contract of request.contracts) {
+    for (const [sourcePath, source] of Object.entries(contract.sources)) {
+      const forkContract: TenderlyContract = {
+        compiler: undefined,
+        contractName: source.name,
+        networks: undefined,
+        source: source.code,
+        sourcePath,
+      };
+      if (
+        forkContract.contractName === contract.contractToVerify ||
+        `${forkContract.sourcePath}:${forkContract.contractName}` === contract.contractToVerify
+      ) {
+        forkContract.networks = contract.networks;
+      }
+      forkRequest.contracts.push(forkContract);
+    }
+  }
+
+  forkRequest.config = {};
+  forkRequest.config.compiler_version = request.contracts[0].compiler?.version;
```

Comment:
If we can remove this kind of hack for parsing/fetching data I would defiantly do it immediately.

Reply:
If you are talking about this particular code block that converts compiler configuration, I'm afraid it is tied to the explanation I gave [here](https://github.com/Tenderly/hardhat-tenderly/pull/101/files#r1057608506).

> When we made multi compiler routes for private and public verification, we didn't make one for fork verification. Hence, the request format for fork verification differs from private and public verification request format. In order to send a request for fork verification, there needs to be conversion of requests, which normally shouldn't be there if the requests were the same format. This is why I am calling `makeForkVerifyContractsRequest` and `_convertToForkRequest` redundant or excessive.
> 
> The effort it takes to change this is to implement multi compiler fork verification. I know there are already .proto request messages and structures made, but I think implementing this is out of scope for this iteration. 😄

Reply:
If there is an intention to remove it in the future I'm totally fine with this. As long as it is going to occur sooner then later 😅